Trusted image storage

ChengyuZhu6 opened the pull request:
Mostly cherry-pick https://github.com/kata-containers/kata-containers/pull/4885 from CCv0

Fixes: #8142

ChengyuZhu6 commented:
Case 1: Failed to pull the large image in the guest in memory:
```
Events:
  Type     Reason   Age                From     Message
  ----     ------   ----               ----     -------
  Normal   Pulled   101s               kubelet  Successfully pulled image "ghcr.io/confidential-containers/test-container:big-size" in 1.221014717s
  Normal   Pulled   83s                kubelet  Successfully pulled image "ghcr.io/confidential-containers/test-container:big-size" in 1.25888686s
  Normal   Pulled   50s                kubelet  Successfully pulled image "ghcr.io/confidential-containers/test-container:big-size" in 1.221221808s
  Warning  Failed   35s (x3 over 85s)  kubelet  Error: failed to create containerd task: failed to create shim task: failed to handle layer: hasher sha256: channel: send failed SendError { .. }: unknown
  Warning  BackOff  21s (x4 over 65s)  kubelet  Back-off restarting failed container
  Normal   Pulling  7s (x4 over 102s)  kubelet  Pulling image "ghcr.io/confidential-containers/test-container:big-size"
  Normal   Created  6s (x4 over 101s)  kubelet  Created container test-container
  Normal   Pulled   6s                 kubelet  Successfully pulled image "ghcr.io/confidential-containers/test-container:big-size" in 1.252684668s
```

ChengyuZhu6 commented:
Case 2: Successfully pull the large image in the guest in the trusted storage:
```
Events:
  Type    Reason                 Age   From               Message
  ----    ------                 ----  ----               -------
  Normal  Scheduled              55s   default-scheduler  Successfully assigned default/large-image-pod to zcy-ubuntu22
  Normal  SuccessfulMountVolume  56s   kubelet            MapVolume.MapPodDevice succeeded for volume "trusted-block-pv" globalMapPath "/var/lib/kubelet/plugins/kubernetes.io~local-volume/volumeDevices/trusted-block-pv"
  Normal  SuccessfulMountVolume  56s   kubelet            MapVolume.MapPodDevice succeeded for volume "trusted-block-pv" volumeMapPath "/var/lib/kubelet/pods/21574dfa-6f74-4520-bf4a-00138330c80f/volumeDevices/kubernetes.io~local-volume"
  Normal  Pulled                 52s   kubelet            Container image "ghcr.io/confidential-containers/test-container:big-size" already present on machine
  Normal  Created                52s   kubelet            Created container app-container
  Normal  Started                1s    kubelet            Started container app-container
```

ChengyuZhu6 commented:
I think the PR is read for review.

stevenhorsman commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -77,8 +55,8 @@ setup() {
     echo "sandbox_id is: $sandbox_id"
-    # With annotation for nydus, only rootfs for pause container can be found on host
-    assert_rootfs_count "$node" "$sandbox_id" "1"
+    # No rootfs can be found on host with guest pull
+    assert_rootfs_count "$node" "$sandbox_id" "0"
```
We could consider removing the `assert_rootfs_count` checks entirely, as we are running with sharedfs=none now?

ChengyuZhu6 replied:
Done.

stevenhorsman commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -9,39 +9,22 @@ load "${BATS_TEST_DIRNAME}/lib.sh"
-        skip "Due to issues related to pull-image integration skip tests for ${KATA_HYPERVISOR}."
-    fi
-
-    if ! is_confidential_runtime_class; then
-        skip "Test not supported for ${KATA_HYPERVISOR}."
-    fi
```
I think we still need this check so that these test don't run on non coco runtimeclasses e.g. qemu-stratovirt?

ChengyuZhu6 replied:
The check will run in both `setup()` and `teardown()` for each test. From my understanding, if the runtime class is not tee’s, should it be skipped directly in the `setup()` phase? Please correct me if I have misunderstood anything.

stevenhorsman replied:
Ah sorry, I missed that this was in the test and we also had the setup and teardown. My mistake!

stevenhorsman commented on tests/integration/kubernetes/runtimeclass_workloads/pod-guest-pull-in-trusted-storage.yaml:
```diff
@@ -0,0 +1,33 @@
+metadata:
+  name: large-image-pod
+spec:
+  runtimeClassName: RUNTIMECLASS
```
This should be `kata` and it gets automatically updated with the correct runtimeclass during testing IIRC

ChengyuZhu6 replied:
Fixed.

stevenhorsman replied:
It still says `runtimeClassName: $RUNTIMECLASS`?

ChengyuZhu6 replied:
We need to replace it with the value of kata-${KATA_HYPERVISOR} using envsubst, similar to how the new_pod_config function does.

stevenhorsman replied:
So if you look inside the other `runtimeclass_workloads` test pod specs you will see we have `runtimeClassName : kata` set. I think it's because in the CI we alias the `kata` runtimeClass to match the platform we are testing.

ChengyuZhu6 replied:
Fixed.

stevenhorsman commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -59,37 +60,59 @@ setup() {
+    pod_config="${pod_config_dir}/pod-guest-pull-in-trusted-storage.yaml"
+    storage_config="${pod_config_dir}/confidential/trusted-storage.yaml"
+    local_device=$(create_loop_device "/tmp/trusted-storage.img")
+    sed -i "s/RUNTIMECLASS/kata-${KATA_HYPERVISOR}/g" $pod_config
```
This isn't needed

ChengyuZhu6 replied:
What do you mean?

stevenhorsman replied:
The RUNTIMECLASS replacement in the pod.yaml is done automatically if it's `kata`, so we don't need the sed in the test code

stevenhorsman commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -59,37 +60,59 @@ setup() {
-    assert_pod_fail "$kata_pod_with_nydus_config"
+    assert_pod_fail "$pod_config"
     assert_logs_contain "$node" kata "$node_start_time" \
 		'context deadline exceeded'
```
Don't we want to increase the create_container timeout and check it fails because the disk is too small, rather than that the pull isn't quick enough?

ChengyuZhu6 replied:
Yes, you're right. I’ve added two tests: one to verify trusted storage, and another to test the configuration of the createContainer timeout.

stevenhorsman commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -16,8 +16,9 @@ setup() {
+    get_pod_config_dir
     unencrypted_image="quay.io/prometheus/busybox:latest"
-    large_image="quay.io/confidential-containers/test-images:largeimage"
+    large_image="ghcr.io/confidential-containers/test-container:big-size"
```
Out of interest, is there any particular reason for ghcr over quay.io?

ChengyuZhu6 replied:
The reason for selecting a different image is that Image `quay.io/confidential-containers/test-images:largeimage` is too large (larger than 2GB) and may exceed the default kubelet timeout of 120s in CI. Even if we set the createContainer timeout value to more than 120 seconds on the kata side, it might still fail. Therefore, I chose a smaller image (size = 956MB) that can be pulled within 120 seconds.

zvonkok replied:
You would need to set the remote request timeout in the kubelet. This is independent of what we set in kata since the context is created in the kubelet.
Do we have the tools in the ci to update kubelet arguments?

zvonkok replied:
The 120s will also break the GPU use case..

ChengyuZhu6 replied:
@zvonkok I think updating the kubelet configuration might be challenging while the test is running. And I think we can focus on the feature that allows configuring the timeout specifically for pulling larger images. If users need to pull an image with a timeout exceeding 120 seconds, they can update the configuration of kubelet offline. While I'm not an expert on this, I would like any advice from the experts if possible.

fitzthum commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -77,167 +56,66 @@ setup() {
-    assert_rootfs_count "$node" "$sandbox_id" "2"
-}
-
-@test "Test we can pull an other unencrypted image outside the guest and then inside the guest successfully" {
```
Maybe we should keep this test?

wainersm replied:
@ChengyuZhu6 , mind to explain why this and `"Test we can pull an unencrypted image inside the guest twice in a row and then outside the guest successfully" ` aren't needed anymore?

fitzthum commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
I think we might want to rename this file. We use guest pulling for all of the coco tests. These tests are more about guest pulling edge cases. 

Also, any particular reason for dropping the tests that you delete in this PR?

ChengyuZhu6 replied:
Currently, we do not need annotation "io.containerd.cri.runtime-handler" to enable guest pull. In other words, we will always pull images in the guest using the runtimeclass which sets snapshotter to nydus. So these removed tests are useless.

stevenhorsman replied:
I think we might need `"io.containerd.cri.runtime-handler"` on some of our tests. IIUC this is needed on Kuberenetes 1.28 and earlier and I think some of our self-hosted runner might have an older level of Kubernetes installed as I tried removing this in the past and guest pull wasn't working. Something to check for in the e2e results.

ChengyuZhu6 replied:
I’m a bit confused. Currently, we have set `shared_fs` to none, which means we can’t share any files or directories with guest through virtio-fs/9p, including the image pulled on the host. As a result, these tests can’t run on CI. Why do we need to keep them?

ChengyuZhu6 replied:
@stevenhorsman I have kept `io.containerd.cri.runtime-handler` in the tests.

fitzthum commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -77,167 +56,66 @@ setup() {
-    kata_pod_with_nydus_config="$(new_pod_config "$unencrypted_image_1" "kata-${KATA_HYPERVISOR}")"
-    set_node "$kata_pod_with_nydus_config" "$node"
-    set_container_command "$kata_pod_with_nydus_config" "0" "sleep" "30"
+    kubectl delete -f $pod_config
```
Do you think it would be better to split the positive and negative cases into two tests? We have done that in some of our other files.

ChengyuZhu6 replied:
Sure.

ChengyuZhu6 replied:
Done.

fitzthum commented on src/agent/src/rpc.rs:
```diff
@@ -236,6 +238,30 @@ impl AgentService {
+                    data_integrity.to_string()
+                );
+
+                Command::new(INIT_TRUSTED_STORAGE)
```
Do you think it could make sense to handle this functionality in the CDH? That might tie-in nicely with when we support persistent storage that we need a key to unlock.

ChengyuZhu6 replied:
That's a good idea. Perhaps it would be better to merge this into CDH when we move forward the secure mount feature?

fitzthum replied:
Yes, although if we're going to move it to the CDH, maybe we should do it that way from the start. I'll leave that for a kata reviewer to think about.

ChengyuZhu6 replied:
Yeah, I have synced with @Xynnn007. He agreed to move the logic to CDH. I’ll handle it in this PR.

Xynnn007 replied:
Yes, we can reuse the script for multiple scenarios a) encrypted storage for extra image storage b) encrypted rbd for container volume. The both could get key a) from KBS or b) generate an ephemeral key.

fitzthum replied:
Ok, I think that's a good idea. It might take a little more work up front, but I think it will make things nicer down the road.

ChengyuZhu6 replied:
https://github.com/confidential-containers/guest-components/issues/540

fitzthum reviewed (commented):
Great work. Thanks for doing the heavy-lifting.

I have a couple questions about our plan for various types of secure storage. 

In CCv0 we had ephemeral block device support (basically this PR) and we also had secure storage driven by the CDH. I think there are actually two different parameters. 1) Whether the storage is ephemeral or persistent. 2) Whether the storage is exposed to the pods or if it is used to pull the containers transparently.

This PR is for ephemeral storage that is not exposed to the pod. The question is if we can make these cases share more code and config than they did in CCv0. For instance, how hard would it be to take an ephemeral device used here and expose it to the pod rather than using it to store the containers? What if we wanted to pull containers into persistent encrypted storage? Maybe this is totally out of scope, but I always felt like the two approaches in CCv0 shoudl have been more connected. wdyt?

ChengyuZhu6 commented on docs/how-to/how-to-pull-images-in-guest-with-kata.md:
```diff
@@ -163,14 +161,152 @@ NAME         READY   STATUS    RESTARTS   AGE
+        - sleep 6000
+      volumeDevices:
+        - devicePath: /dev/trusted_store
+          name: trusted-storage
```
@fitzthum The device is exposed to the pod instead container. In this example, only one image is pulled. Therefore, if multiple containers with different images are started in a pod, we only need to define the volume device in one of them. All images would be downloaded in the trusted storage.

fitzthum replied:
Yes, it is at the scope of the pod, but I am saying that in normal kubernetes the behavior of this config would be different. Normal k8s would treat this volumeDevice just like any other volumeDevice, but for us the name field triggers different behavior. This is what we should make sure to document.

ChengyuZhu6 commented on docs/how-to/how-to-pull-images-in-guest-with-kata.md:
````diff
@@ -163,14 +161,152 @@ NAME         READY   STATUS    RESTARTS   AGE
+volumeBindingMode: WaitForFirstConsumer
+```
+
+3. Create PersistentVolume
````
>Also, do you think it's weird to be using persistent volumes for something that is essentially single-use? Are users expected to create a new volume for each pod or can it be reused? Maybe we should have some info about that.

Yeah, currently, we create a pv for tests. Previous works on the CCv0 branch involved creating PVs using open-local, which were subsequently managed by the CSI driver for cleanup. Without a third-party CSI driver, k8s cannot dynamically clean up the created PVs, so we need manual cleanup. To minimize additional components and reduce complexity in CI, I recommend avoiding the installation of extra CSI drivers.

ChengyuZhu6 replied:
>k8s does have ephemeral volumes as well. Do you think those would work here?

From my understanding, k8s ephemeral volumes can be effectively used for mounting as file systems or directories without requiring any third-party CSI drivers. However, if we need to use them as block volumes, we still require a third-party CSI driver that supports block devices. Please correct me if I have misunderstood anything.

ChengyuZhu6 replied:
Update info: I attempted to use one of k8s’ ephemeral storage types, `emptyDir`, to achieve trusted storage. However, it did not work for the reason:
The `emptyDir` type can only be used as a tmpfs mounted to the guest, and its size is influenced by the RAM allocated to the guest.

ChengyuZhu6 commented on tools/osbuilder/rootfs-builder/scripts/init_trusted_storage.sh:
```diff
@@ -0,0 +1,130 @@
+#!/bin/bash
+#
+# Copyright (c) 2022 Intel Corporation
```
Fixed.

stevenhorsman replied:
This looks to be copy-pasted directly from the `CCv0` version, so the original copyright date would be correct

ChengyuZhu6 replied:
Reverted.

ChengyuZhu6 commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -246,4 +124,6 @@ teardown() {
     kubectl describe pod "$pod_name"
     k8s_delete_all_pods_if_any_exists || true
+    kubectl delete -f "${pod_config_dir}/pod-guest-pull-in-trusted-storage.yaml" || true
+    kubectl delete -f "${pod_config_dir}/confidential/trusted-storage.yaml" || true
```
Yes

c3d commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -77,167 +56,66 @@ setup() {
-    # The test case will be PASS after we use containerd 2.0 with 'image pull per runtime class' feature:
-    # https://github.com/containerd/containerd/issues/9377
-    assert_rootfs_count "$node" "$sandbox_id" "1"
+    add_allow_all_policy_to_yaml "$pod_config"
```
@fitzthum The status of OPA policies a good question, but I think that you'd give it more visibility outside of this specific PR (e.g. on Slack).

On this PR specifically, the question of what is the correct way to do it in a test is a really good one. At the very least, a comment would be welcome.

ChengyuZhu6 commented:
After discussion, we decided to have CDH handle the logic for encrypting the block device. We need to wait for the PR https://github.com/confidential-containers/guest-components/pull/617 to merge.

ChengyuZhu6 commented:
Mark the PR as draft.

ChengyuZhu6 commented on src/agent/src/config.rs:
```diff
@@ -29,6 +29,7 @@ const UNIFIED_CGROUP_HIERARCHY_OPTION: &str = "systemd.unified_cgroup_hierarchy"
 const CONFIG_FILE: &str = "agent.config_file";
 const GUEST_COMPONENTS_REST_API_OPTION: &str = "agent.guest_components_rest_api";
 const GUEST_COMPONENTS_PROCS_OPTION: &str = "agent.guest_components_procs";
+const DATA_INTEGRITY_OPTION: &str = "agent.data_integrity";
```
@c3d This information is based on commit https://github.com/kata-containers/kata-containers/pull/4885/commits/b4c4ec66d067edcc42c5560a85728dd5d39ebeff. Once commit https://github.com/arronwy/kata-containers/commit/4906228701b0a71ef8d2e941ce7cca2b8748ad40 is added to optimize the integrity device creation, the actual overhead may be reduced.

ChengyuZhu6 commented:
Currently, I attempted to update `coco-guest-components` to the commit https://github.com/confidential-containers/guest-components/commit/0d8146321c3de023f0f7f40e47fc0f860133dfc7 that supports trusted storage. However, all related coco CIs failed due to attestation agent initialization error. I have created an issue in confidential-containers to track this: https://github.com/confidential-containers/guest-components/issues/626. @Xynnn007 has quickly addressed this with a PR: https://github.com/confidential-containers/guest-components/pull/627. Thanks for the help, @Xynnn007!

sprt commented on tools/osbuilder/rootfs-builder/scripts/init_trusted_storage.sh:
```diff
@@ -0,0 +1,149 @@
+	if [ "$is_encrypted" == "false" ]; then
+
+		if [ "$data_integrity" == "false" ]; then
+			echo "YES" | cryptsetup luksFormat --type luks2 "$device_path" --sector-size 4096 \
```
I believe we can use the `-q/--batch-mode` flag here instead of passing `YES`.

ChengyuZhu6 replied:
Nice catch. Thanks!

ChengyuZhu6 replied:
Fixed.

fitzthum commented on src/agent/src/config.rs:
```diff
@@ -110,6 +111,7 @@ pub struct AgentConfig {
     pub guest_components_procs: GuestComponentsProcs,
     #[cfg(feature = "guest-pull")]
     pub image_registry_auth: String,
+    pub data_integrity: bool,
```
`data_integrity` is a pretty generic term. Maybe this should be something more specific to secure storage?

ChengyuZhu6 replied:
I have renamed `data_integrity` to `secure_storage_integrity`.

fitzthum reviewed (commented):
Looks great. Two small notes.

ChengyuZhu6 commented on docs/how-to/how-to-pull-images-in-guest-with-kata.md:
```diff
@@ -163,14 +161,152 @@ NAME         READY   STATUS    RESTARTS   AGE
+
+## Run pod in kata containers with pulling large image in guest
+
+Currently, the image pulled in the guest will be downloaded and unpacked in the `/run/kata-containers/image` directory. However, by default, systemd allocates 10% of the available physical RAM to the `/run` directory using a `tmpfs` filesystem. As we all know, memory is valuable, especially for confidential containers. This means that if we run a kata container with the default configuration (whe ...
```
Ah, good point. I’ve updated the description.

ChengyuZhu6 replied:
And I've updated the example.

ChengyuZhu6 commented:
rebase to main to fix the issue with the deployment of kbs.

ChengyuZhu6 commented:
@arronwy  @fitzthum @zvonkok @Xynnn007 @stevenhorsman @sprt @c3d The tests about trusted storage have passed on some TEE and non-TEE platforms. BTW, this PR is unrelated to TEE. I will track the failed tests, but I think this PR is ready for another review. I have addressed the issues you mentioned earlier and I appreciate it if you could review it again.. Thank you.

fitzthum commented:
Code looks good to me.

The nontee test is failing https://github.com/kata-containers/kata-containers/actions/runs/10193791423/job/28206001226?pr=9999#step:14:1714

ChengyuZhu6 commented:
> Code looks good to me.
> 
> The nontee test is failing https://github.com/kata-containers/kata-containers/actions/runs/10193791423/job/28206001226?pr=9999#step:14:1714

Yes, it might be due to the lack of available loop devices on the node. I’ll need the CI maintainer’s help with this.

ChengyuZhu6 commented:
The job on s390x failed because the image `quay.io/confidential-containers/test-images:largeimage` timed out while being pulled in the guest. The image size is 2.15GB.

fitzthum commented:
> Yes, it might be due to the lack of available loop devices on the node. I’ll need the CI maintainer’s help with this.

I think these tests run in the AKS environment, which might make things a little tricky.

sprt commented on tests/integration/kubernetes/confidential_common.sh:
```diff
@@ -82,3 +82,28 @@ function is_confidential_hardware() {
+
+	sudo dd if=/dev/zero of=$loop_file bs=1M count=2500
+	sudo losetup -fP $loop_file >/dev/null 2>&1
+	local device=$(sudo losetup -j $loop_file | awk -F'[: ]' '{print $1}')
```
You have to use `exec_host()` here, otherwise the loop device will be created on the runner machine, rather than the host node.

ChengyuZhu6 replied:
Thanks for your comment @sprt. Let's change the code and see if the test passes on coco-dev CI.

ChengyuZhu6 commented on tools/osbuilder/rootfs-builder/rootfs.sh:
```diff
@@ -750,6 +750,11 @@ EOF
 
+	if [ "${MEASURED_ROOTFS}" == "yes" ]; then
+		info "Install init_trusted_storage script"
+		install -o root -g root -m 0500 "${script_dir}/scripts/init_trusted_storage.sh" "${ROOTFS_DIR}/usr/local/bin/luks-encrypt-storage"
```
>This is installed for the CDH, correct?

Yes. Currently, CDH is not responsible for installing the script, so we need to install it in kata.

> As in, only the CDH is expected to call this script?
Yes, that's correct.

ChengyuZhu6 commented:
>Agreed, I think we should clearly document this. It's also confusing to call this feature "trusted storage" as the actual scope is more constrained - I'd call it "trusted image storage".

Sure. Let's rename it to `trusted image storage`.

ChengyuZhu6 commented on src/agent/src/rpc.rs:
```diff
@@ -243,6 +245,37 @@ impl AgentService {
+                            ("encryptType".to_string(), "LUKS".to_string()),
+                            ("dataIntegrity".to_string(), secure_storage_integrity),
+                        ]);
+                        cdh.secure_mount("BlockDevice", &options, vec![], KATA_IMAGE_WORK_DIR)
```
Yes, and for the rootfs-confidential image, the size of `/run` as tmpfs is 50% of the memory.

ChengyuZhu6 commented:
rebase to main only.

ChengyuZhu6 commented:
Because the issue https://github.com/kata-containers/kata-containers/issues/10133 by using `exec_host` to manage loop device. So the tests with trusted storage will be only running in baremetal machines.

sprt reviewed (approved):
LGTM! Just 2 nits. 🙂

ChengyuZhu6 commented on src/agent/src/rpc.rs:
```diff
@@ -106,6 +107,7 @@ use kata_types::k8s;
 
 pub const CONTAINER_BASE: &str = "/run/kata-containers";
 const MODPROBE_PATH: &str = "/sbin/modprobe";
+const TRUSTED_STORAGE_DEVICE: &str = "/dev/trusted_store";
```
Fixed.

ChengyuZhu6 commented on src/agent/src/config.rs:
```diff
@@ -31,6 +31,7 @@ const GUEST_COMPONENTS_REST_API_OPTION: &str = "agent.guest_components_rest_api"
 const GUEST_COMPONENTS_PROCS_OPTION: &str = "agent.guest_components_procs";
 #[cfg(feature = "guest-pull")]
 const IMAGE_REGISTRY_AUTH_OPTION: &str = "agent.image_registry_auth";
+const SECURE_STORAGE_INTEGRITY_OPTION: &str = "agent.secure_storage_integrity";
```
This option is used for secure storage，not only for image storage. I recommend to keep this name.

wainersm replied:
Hi @ChengyuZhu6 !

Could you please update https://github.com/kata-containers/kata-containers/blob/main/src/agent/README.md#agent-options ?

ChengyuZhu6 replied:
Sure!

ChengyuZhu6 replied:
Done.

wainersm replied:
thx!

wainersm commented on src/agent/src/rpc.rs:
```diff
@@ -106,6 +106,8 @@ use kata_types::k8s;
 pub const CONTAINER_BASE: &str = "/run/kata-containers";
 const MODPROBE_PATH: &str = "/sbin/modprobe";
+const INIT_TRUSTED_STORAGE: &str = "/usr/bin/kata-init-trusted-storage";
+const TRUSTED_IMAGE_STORAGE_DEVICE: &str = "/dev/trusted_store";
```
to keep it coherent, shouldn't it be `TRUSTED_STORAGE_DEVICE` (without IMAGE)?

ChengyuZhu6 replied:
This device is used for image storage, but the script `kata-init-trusted-storage` serves additional scenarios beyond just  trusted image storage.

wainersm commented on src/agent/src/rpc.rs:
```diff
@@ -243,6 +245,32 @@ impl AgentService {
+            .ok_or_else(|| anyhow!("Spec didn't contain linux field"))?;
+        if let Some(devices) = linux.devices() {
+            for specdev in devices.iter() {
+                let dev_major_minor = format!("{}:{}", specdev.major(), specdev.minor());
```
micro-optimization: move this assignment to the `if specdev.path().....` block below.

ChengyuZhu6 replied:
Fixed.

wainersm commented on src/agent/src/rpc.rs:
```diff
@@ -243,6 +245,32 @@ impl AgentService {
+                        .args([dev_major_minor.as_str(), &secure_storage_integrity])
+                        .output()
+                        .expect("Failed to initialize confidential storage");
+                }
```
`break` the `for` loop once trusted storage is initialized ;)

ChengyuZhu6 replied:
Done.

wainersm commented on src/agent/src/rpc.rs:
```diff
@@ -243,6 +245,32 @@ impl AgentService {
+                    Command::new(INIT_TRUSTED_STORAGE)
+                        .args([dev_major_minor.as_str(), &secure_storage_integrity])
+                        .output()
+                        .expect("Failed to initialize confidential storage");
```
suggestion: s/confidential/trusted/

ChengyuZhu6 replied:
Done.

wainersm commented on tools/osbuilder/rootfs-builder/scripts/init_trusted_storage.sh:
```diff
@@ -0,0 +1,145 @@
+#!/bin/bash
```
Is this script a copy of https://github.com/confidential-containers/guest-components/blob/main/confidential-data-hub/storage/scripts/luks-encrypt-storage ? if so then I think we should fetch the file from guest-components otherwise it can get easily out of sync and introduce regressions hard to find.

One question that I have is if any plan to replace the script by a rust app or even a function in guest-components.

ChengyuZhu6 replied:
@wainersm I think we can enhance this further in our follow-up work. WDYT?

ChengyuZhu6 replied:
>Is this script a copy of https://github.com/confidential-containers/guest-components/blob/main/confidential-data-hub/storage/scripts/luks-encrypt-storage ? if so then I think we should fetch the file from guest-components otherwise it can get easily out of sync and introduce regressions hard to find.

Yes, I agree with you. Although the PR includes many commits, I recommend addressing the work in the next PR. WDYT?

wainersm replied:
Hi @ChengyuZhu6 ! I'm okay to address these comments in a follow on PR

wainersm replied:
@ChengyuZhu6 created an issue to track that improvement: https://github.com/confidential-containers/guest-components/issues/670

ChengyuZhu6 commented:
rebase to main only

wainersm commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -72,13 +55,6 @@ setup() {
     k8s_create_pod "$kata_pod_with_nydus_config"
-    echo "Kata pod test-e2e with nydus annotation is running"
-
-    echo "Checking the image was pulled in the guest"
```
Hi @ChengyuZhu6 !

I didn't understand why you removed this check. How do we ensure it was really pulled by the guest?

ChengyuZhu6 replied:
Hi @wainersm! In the coco CI job, we're setting `shared_fs` to `none`, as seen in the Makefile: https://github.com/kata-containers/kata-containers/blob/91084058ae620d7c3c2053cdac2de0445824fbc2/src/runtime/Makefile#L248-L252. This change implies that all kubernetes tests will now pull images in the guest. Consequently, there's no longer a need to check.

wainersm replied:
My understanding of `shared_fs=none` is that even not mounting filesystems, it will be copying files over the guest (e.g. config maps). I don't know if it will attempt to copy images...

Anyway, afaik we don't have any test to test `shared_fs=none` really works. Having that checking on this test case will ensure that images are pulling from the guest as intended. Is it too much work keep this checking?

wainersm commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -57,42 +61,66 @@ setup() {
+    # However, by default, systemd allocates 50% of the available physical RAM to the `/run` directory using a `tmpfs` filesystem. 
+    # It means that if we run a kata container with the default configuration (where the default memory assigned for a VM is 2048 MiB), 
+    # `/run` would be allocated around 1024 MiB. Consequently, we can only pull images up to 1024 MiB in the guest. 
+    # However, the unpacked size of image "rust:latest" is 1.41GB. 
```
why don't keep using quay.io/confidential-containers/test-images:largeimage which is larger than 1.41GB, so will fail as well?

ChengyuZhu6 replied:
Because the default pull timeout is 60s and it would be timeout if we pull `quay.io/confidential-containers/test-images:largeimage` in the case `Test we can pull an image inside the guest using trusted storage`.

wainersm replied:
ah, ok. so it should large enough to get pulled before the deadline. Thanks for the explanation!

wainersm commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -16,8 +16,12 @@ setup() {
+    get_pod_config_dir
     unencrypted_image="quay.io/prometheus/busybox:latest"
-    large_image="quay.io/confidential-containers/test-images:largeimage"
+    image_pulled_time_less_than_default_time="rust:latest" # unpacked size: 1.41GB
```
`rust:latest` will be pulled from docker.io so changes are CI will hit pull rate limit issue. We better find a large image from quay.io, or maybe this image is not needed (see my next comment)

ChengyuZhu6 replied:
Yes, I agree. But I’ve been searching for an appropriate and dependable image from quay.io that is larger than 1GB yet smaller than 2GB, but I haven’t found one that meets these criteria yet. Do you have any advice?

ChengyuZhu6 replied:
@wainersm  I have replaced "rust:latest" with "ghcr.io/confidential-containers/test-container:rust-1.79.0". Thanks @Xynnn007's help!

wainersm commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -57,42 +61,66 @@ setup() {
+    local_device=$(create_loop_device)
+    LOCAL_DEVICE="$local_device" NODE_NAME="$node" envsubst < "$storage_config_template" > "$storage_config"
 
-    # Set CreateContainerRequest timeout in the annotation to pull large image in guest
```
why timeout is not needed anymore? Don't we have the risk of hitting it?

wainersm replied:
ok, next commit introduce the tests with timeout.

wainersm commented on tests/integration/kubernetes/confidential_common.sh:
```diff
@@ -82,3 +82,28 @@ function is_confidential_hardware() {
+
+function create_loop_device(){
+	local loop_file="${1:-/tmp/trusted-image-storage.img}"
+	cleanup_loop_device
```
Should pass `loop_file` to `cleanup_loop_device`:  `cleanup_loop_device "$loop_file"`

ChengyuZhu6 replied:
Fixed.

wainersm commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -102,6 +130,10 @@ teardown() {
-    kubectl describe pod "$pod_name"
+    kubectl describe pods
     k8s_delete_all_pods_if_any_exists || true
+    kubectl delete pvc trusted-pvc || true
```
Please run these `kubectl delete` with `--ignore-not-found` (and without `|| true`). For example, `kubectl delete --ignore-not-found pvc trusted-pvc`.

So avoiding nasty messages like:
```
2024-08-13T09:03:23.7388751Z # Error from server (NotFound): persistentvolumeclaims "trusted-pvc" not found
2024-08-13T09:03:23.7390598Z # Error from server (NotFound): persistentvolumes "trusted-block-pv" not found
2024-08-13T09:03:23.7392602Z # Error from server (NotFound): storageclasses.storage.k8s.io "local-storage" not found
```

ps: I know that many tests don't pass that parameter, but it's not always too late to learn new habits :)

ChengyuZhu6 replied:
Good point. Thanks for your suggestion @wainersm !

ChengyuZhu6 replied:
Fixed.

wainersm commented on tests/integration/kubernetes/k8s-guest-pull-image.bats:
```diff
@@ -72,178 +59,175 @@ setup() {
+    # The pod should be failed because the default timeout of CreateContainerRequest is 60s
+    assert_pod_fail "$pod_config"
+    assert_logs_contain "$node" kata "$node_start_time" \
+		'context deadline exceeded'
```
Unfortunately this error message is pretty generic and can indicate any error. As we have a test that increases the timeout and it should pass then I think we are fine.

But if we could improve the message (not for this PR)...the problem is that it cannot properly propagate errors within the guest?

ChengyuZhu6 replied:
Yes, I agree. I’ll address it in the upcoming PR and keep you updated.

wainersm commented:
Hi @ChengyuZhu6 !

First, thanks for yet another great work! I don't have any further comments to this PR, the improvements can be done on future PRs IMHO. I suppose you will update this PR once more, then we can check if tests still pass.

ChengyuZhu6 commented:
> Hi @ChengyuZhu6 !
> 
> First, thanks for yet another great work! I don't have any further comments to this PR, the improvements can be done on future PRs IMHO. I suppose you will update this PR once more, then we can check if tests still pass.

@wainersm Thanks for your comments!

wainersm commented:
Hi @ChengyuZhu6 !

The tests passed on SNP and SEV
TDX failed the job but on unrelated tests, k8s-guest-pull-image.bats passed.
zvsi failed even before running the tests but it leverages qemu-coco-dev, so it should skip k8s-guest-pull-image.bats tests.

wainersm reviewed (approved):
Hi @ChengyuZhu6 !

We know that there are improvements to be done, we've discussed on this PR and reached on an agreement (at least between you and me :)

This is passing tests on all TEE platforms!

So IMO it's ready to be merged. Thanks again for this amazing work!

fitzthum reviewed (approved):
Great work Chengyu. Thanks!

arronwy reviewed (approved):
Thanks @ChengyuZhu6 LGTM!
